activerecord/lib/active_record/attribute_mutation_tracker.rb

```diff
@@ -31,7 +31,7 @@ def change_to_attribute(attr_name)
     end
 
     def any_changes?
-      attr_names.any? { |attr| changed?(attr) }
+      attr_names.any? { |attr| changed?(attr) } || forced_changes.any?
```

Comment:
By "virtual attribute" are you referring to an `attr_accessor`? Those are not intended to work with dirty.